libraries/AP_Motors/AP_MotorsMatrix_6DoF_Scripting.cpp

```diff
@@ -0,0 +1,325 @@
+    const float compensation_gain = get_compensation_gain(); // compensation for battery voltage and altitude
+    roll_thrust = (_roll_in + _roll_in_ff) * compensation_gain;
+    pitch_thrust = (_pitch_in + _pitch_in_ff) * compensation_gain;
+    yaw_thrust = (_yaw_in + _yaw_in_ff) * compensation_gain;
+    throttle_thrust = get_throttle() * compensation_gain;
+
+    // scale horizontal thrust with throttle, this mimics a normal copter
+    // so we don't break the lean angle proportional acceleration assumption made by the position controller
+    forward_thrust = get_forward() * throttle_thrust;
+    right_thrust = get_lateral() * throttle_thrust;
+
+
+    // set throttle limit flags
+    if (throttle_thrust <= 0) {
+        throttle_thrust = 0;
+        // we cant thrust down, the vehicle can do it, but it would break a lot of assumptions further up the control stack
+        // 1G decent probably plenty anyway....
+        limit.throttle_lower = true;
+    }
+    if (throttle_thrust >= 1) {
+        throttle_thrust = 1;
+        limit.throttle_upper = true;
+    }
+
+    // rotate the thrust into bodyframe
+    Matrix3f rot;
+    Vector3f thrust_vec;
+    rot.from_euler312(_roll_offset, _pitch_offset, 0.0f);
+
+
+    /*
+        upwards thrust, independent of orentaiton
+    */
+    thrust_vec.x = 0.0f;
+    thrust_vec.y = 0.0f;
+    thrust_vec.z = throttle_thrust;
+    thrust_vec = rot * thrust_vec;
+    for (i = 0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (motor_enabled[i]) {
+            _thrust_rpyt_out[i] =  thrust_vec.x * _forward_factor[i];
+            _thrust_rpyt_out[i] += thrust_vec.y * _right_factor[i];
+            _thrust_rpyt_out[i] += thrust_vec.z * _throttle_factor[i];
+
+            if (fabsf(_thrust_rpyt_out[i]) >= 1) {
+                // if we hit this the mixer is probably scaled incorrectly
+                limit.throttle_upper = true;
+            }
+            _thrust_rpyt_out[i] = constrain_float(_thrust_rpyt_out[i],-1.0f,1.0f);
+        }
+    }
+
+
+    /*
+        rotations
+    */
+    float rpy_ratio = 1.0f;  // scale factor, output will be scaled by this ratio so it can all fit evenly
+    float thrust[AP_MOTORS_MAX_NUM_MOTORS];
+    for (i = 0; i < AP_MOTORS_MAX_NUM_MOTORS; i++) {
+        if (motor_enabled[i]) {
+            thrust[i] =  roll_thrust * _roll_factor[i];
+            thrust[i] += pitch_thrust * _pitch_factor[i];
+            thrust[i] += yaw_thrust * _yaw_factor[i];
+            float total_thrust = _thrust_rpyt_out[i] + thrust[i];
+            // control input will be limited by motor range
+            if (total_thrust > 1.0f) {
+                rpy_ratio = MIN(rpy_ratio,(1.0f - _thrust_rpyt_out[i]) / thrust[i]);
+            } else if (total_thrust < -1.0f) {
+                rpy_ratio = MIN(rpy_ratio,(-1.0f -_thrust_rpyt_out[i]) / thrust[i]);
+            }
+        }
+    }
+
+    // set limit flags if output is being scaled
+    if (rpy_ratio < 1) {
+        limit.roll = 1;
```

Comment:
normally we use "true" and "false" although technically they are actually uint8_t bits.. but let's use true/false anyway.

BTW, are these limits ever unset?